grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -1,14 +1,112 @@
 #!/usr/bin/env python
 # Lint as: python3
 """A module with API methods related to the GRR metadata."""
-from typing import Optional
+import json
+import inspect
+
+from urllib import parse as urlparse
+from typing import Optional, cast
+from typing import Type, Any, Union, Tuple, List, Set, Dict, Callable
+
+from google.protobuf.descriptor import Descriptor
+from google.protobuf.descriptor import EnumDescriptor
+from google.protobuf.descriptor import FieldDescriptor
 
 from grr_response_core import version
 from grr_response_core.lib.rdfvalues import structs as rdf_structs
+from grr_response_core.lib.rdfvalues import proto2 as protobuf2
 from grr_response_proto.api import metadata_pb2
 from grr_response_server import access_control
 from grr_response_server.gui import api_call_handler_base
 
+# Type aliases used throughout the metadata module.
+SchemaReference = Dict[str, str]
+PrimitiveSchema = Dict[str, str]
+ArraySchema = Dict[str, Union[str, PrimitiveSchema, SchemaReference]]
+EnumSchema = Dict[str, Union[str, Tuple[str, ...]]]
+MessageSchema = Dict[
+  str, Union[str,
+             Dict[str, Union[SchemaReference, PrimitiveSchema, ArraySchema]]]
+]
+Schema = Union[PrimitiveSchema, EnumSchema, MessageSchema, ArraySchema]
+PrimitiveDescription = Dict[str, Union[str, PrimitiveSchema]]
+
+# Follows https://developers.google.com/protocol-buffers/docs/proto3#json as a
+# base, but whenever https://swagger.io/specification/#data-types provides a
+# more specific description of the same type, the OpenAPI Specification
+# version is used (noted by "OAS (type, format)" comments), with the exception
+# of int64 which uses "string" as a type.
```

Comment:
Instead of putting links inline, I recommend using reference style:

```
Lorem ipsum dolor sit amet, consectetur adipiscing elit [1], sed do eiusmod 
tempor incididunt ut labore et dolore magna aliqua [2].

[1]: http://google.com/
[2]: http://github.com/
```

Also, `int64` seems to be unquoted, whereas `string` is quoted but both refer to type names. I would use Markdown backticks for them (and in general, I tend to use Markdown for all code comments).

Reply:
Thank you, it is more readable now. I used Markdown backticks for all references to code / OpenAPI Specification symbols in comments.